### ISSUE TITLE
Radius varies with the number of slices

### ISSUE THREAD

**Issue by Bikeman868** [psychobolt/react-pie-menu#11]
I am experiencing an issue where the pie menu is double its normal size when there are very few slices. In this case the slices are also clipped, presumably because they overflow the pie menu container.

This is the code for my component:

```javascript
import React, { useState, useEffect } from 'react';
import { Typography } from '@material-ui/core';
import PieMenu, { Slice } from 'react-pie-menu';
import { ThemeProvider, css } from 'styled-components';
import OctopediaService from '~services/OctopediaService';

// See https://github.com/psychobolt/react-pie-menu

export const PopupMenu = ({ selected, onProperties, onDrillDown, onExpandNode, onHideNode, onChangeView }) => {
  const [views, setViews] = useState(null);

  // This effect gets a list of views relating to the node
  useEffect(() => {
    if (!selected || !selected.graphNode || !selected.graphNode.tags || !onChangeView) {
      setViews(null);
      return;
    }
    const promises = [];
    const relatedViews = [];
    const octopediaService = new OctopediaService();
    selected.graphNode.tags.forEach((tag) => {
      promises.push(
        octopediaService.getViews(tag).then((viewsByCategory) => {
          viewsByCategory.forEach((category) => {
            category.views.forEach((categoryView) => {
              categoryView.nodeType = tag;
              relatedViews.push(categoryView);
            });
          });
        }),
      );
    });
    Promise.all(promises).then(() => {
      setViews(relatedViews);
    });
  }, [onChangeView, selected]);

  const handleProperties = () => {
    if (onProperties) onProperties(selected.graphNode);
  };

  const handleExpand = () => {
    if (onExpandNode) onExpandNode(selected.graphNode);
  };

  const handleHide = () => {
    if (onHideNode) onHideNode(selected.graphNode);
  };

  const handleDrillDown = () => {
    if (onDrillDown) onDrillDown(selected.graphNode);
  };

  const handleRelated = (view) => {
    if (onChangeView) onChangeView(selected.graphNode, view);
  };

  const showExpand = !!onExpandNode;
  const showHide = !!onHideNode;
  const showDrillDown = !!onDrillDown;

  const theme = {
    pieMenu: {
      item: css`
        overflow: hidden;
      `,
      container: css`
        border: 4px solid white;
        background: rgba(255, 255, 255, 0.85);
        overflow: visible;
      `,
      center: css`
        background: rgba(0, 0, 0, 0.1);
      `,
    },
    slice: {
      item: css`
        overflow: visible;
      `,
      container: css`
        cursor: pointer;
        overflow: visible;
        border: 1px solid lightgrey;
        background: rgba(0, 0, 0, 0.02);
        &[kind='action'] {
          background: rgba(128, 255, 128, 0.1);
        }
        &[kind='nav'] {
          background: rgba(128, 128, 255, 0.08);
        }
        &:hover,
        &[active='true'],
        &[_highlight='true'] {
          background: rgba(255, 255, 255, 0.85);
        }
      `,
      contentContainer: css``,
      content: css`
        padding-left: 60px;
        padding-right: 60px;
      `,
    },
  };

  return (
    selected &&
    views && (
      <ThemeProvider theme={theme}>
        <PieMenu radius="120px" centerRadius="30px" centerX={selected.x + 'px'} centerY={selected.y + 'px'}>
          <Slice onSelect={handleProperties} contentHeight="5em" attrs={{ kind: 'action' }}>
            <Typography variant="caption" color="primary">
              {selected.graphNode.name} properties
            </Typography>
          </Slice>
          {showExpand && (
            <Slice onSelect={handleExpand} contentHeight="5em" attrs={{ kind: 'action' }}>
              <Typography variant="caption" color="primary">
                Expand {selected.graphNode.name}
              </Typography>
            </Slice>
          )}
          {showHide && (
            <Slice onSelect={handleHide} contentHeight="5em" attrs={{ kind: 'action' }}>
              <Typography variant="caption" color="primary">
                Hide {selected.graphNode.name}
              </Typography>
            </Slice>
          )}
          {showDrillDown && (
            <Slice onSelect={handleDrillDown} contentHeight="5em" attrs={{ kind: 'action' }}>
              <Typography variant="caption" color="primary">
                Focus on {selected.graphNode.name}
              </Typography>
            </Slice>
          )}
          {views.map((view) => (
            <Slice
              key={view.id}
              contentHeight="5em"
              attrs={{ kind: 'nav' }}
              onSelect={() => {
                handleRelated(view);
              }}>
              <Typography variant="caption" color="textPrimary">
                {view.name}
              </Typography>
            </Slice>
          ))}
        </PieMenu>
      </ThemeProvider>
    )
  );
};
```

**Comment by Bikeman868** [comment]
Further experimentation suggests that problem manifests with anything less than 4 slices.

**Comment by psychobolt** [comment]
Hi, you should keep a radial constraint on the background styles e.g.:

from

```css
  background: rgba(128, 255, 128, 0.1);
```

to

```js
`
   background: radial-gradient(transparent ${({ centerRadius }) => `${centerRadius}, rgba(128, 255, 128, 0.1); ${centerRadius}`});
`
```

See example: 
https://github.com/psychobolt/react-pie-menu/blob/5a16a54e195d6d015c2a8404984db40cc8b2a1f6/stories/PreferenceSelector/PreferenceSelector.style.js#L24

**Comment by Bikeman868** [comment]
Changing the background to a redial gradient makes the pie slices not paint in the center of the pie, but otherwise the behavior is the same. The menu is still double the normal size with less than 4 slices.

This is my revised code:
```javascript
import React, { useState, useEffect } from 'react';
import { Typography } from '@material-ui/core';
import PieMenu, { Slice } from 'react-pie-menu';
import { ThemeProvider, css } from 'styled-components';
import DataCache from '~services/DataCache';

// See https://github.com/psychobolt/react-pie-menu

export const PopupMenu = ({ selected, onProperties, onDrillDown, onExpandNode, onHideNode, onChangeView }) => {
  const [views, setViews] = useState(null);

  // This effect gets a list of views relating to the node
  useEffect(() => {
    if (!selected || !selected.graphNode || !selected.graphNode.tags || !onChangeView) {
      setViews(null);
      return;
    }
    const promises = [];
    const relatedViews = [];
    const dataCache = new DataCache();
    selected.graphNode.tags.forEach((tag) => {
      promises.push(
        dataCache.getViews(tag).then((viewsByCategory) => {
          viewsByCategory.forEach((category) => {
            category.views.forEach((categoryView) => {
              categoryView.nodeType = tag;
              relatedViews.push(categoryView);
            });
          });
        }),
      );
    });
    Promise.all(promises).then(() => {
      setViews(relatedViews);
    });
  }, [onChangeView, selected]);

  const handleProperties = () => {
    if (onProperties) onProperties(selected.graphNode);
  };

  const handleExpand = () => {
    if (onExpandNode) onExpandNode(selected.graphNode);
  };

  const handleHide = () => {
    if (onHideNode) onHideNode(selected.graphNode);
  };

  const handleDrillDown = () => {
    if (onDrillDown) onDrillDown(selected.graphNode);
  };

  const handleRelated = (view) => {
    if (onChangeView) onChangeView(selected.graphNode, view);
  };

  const showExpand = !!onExpandNode;
  const showHide = !!onHideNode;
  const showDrillDown = !!onDrillDown;

  const theme = {
    pieMenu: {
      item: css`
        overflow: hidden;
      `,
      container: css`
        border: 4px solid white;
        background: radial-gradient(transparent ${({ centerRadius }) => `${centerRadius}, rgba(255, 255, 255, 0.85) ${centerRadius}`});
        overflow: visible;
      `,
      center: css`
        background: transparent;
      `,
    },
    slice: {
      item: css`
        overflow: visible;
      `,
      container: css`
        cursor: pointer;
        overflow: visible;
        border: 1px solid lightgrey;
        &[kind='action'] {
          background: radial-gradient(transparent ${({ centerRadius }) => `${centerRadius}, rgba(128, 255, 128, 0.1) ${centerRadius}`});
        }
        &[kind='nav'] {
          background: radial-gradient(transparent ${({ centerRadius }) => `${centerRadius}, rgba(128, 128, 255, 0.08) ${centerRadius}`});
        }
        &:hover,
        &[active='true'],
        &[_highlight='true'] {
          background: radial-gradient(transparent ${({ centerRadius }) => `${centerRadius}, rgba(255, 255, 255, 0.85) ${centerRadius}`});
        }
      `,
      contentContainer: css``,
      content: css`
        padding-left: 60px;
        padding-right: 60px;
      `,
    },
  };

  return (
    selected && (
      <ThemeProvider theme={theme}>
        <PieMenu radius="120px" centerRadius="30px" centerX={selected.x + 'px'} centerY={selected.y + 'px'}>
          <Slice onSelect={handleProperties} contentHeight="5em" attrs={{ kind: 'action' }}>
            <Typography variant="caption" color="primary">
              {selected.graphNode.name} properties
            </Typography>
          </Slice>
          {showExpand && (
            <Slice onSelect={handleExpand} contentHeight="5em" attrs={{ kind: 'action' }}>
              <Typography variant="caption" color="primary">
                Expand {selected.graphNode.name}
              </Typography>
            </Slice>
          )}
          {showHide && (
            <Slice onSelect={handleHide} contentHeight="5em" attrs={{ kind: 'action' }}>
              <Typography variant="caption" color="primary">
                Hide {selected.graphNode.name}
              </Typography>
            </Slice>
          )}
          {showDrillDown && (
            <Slice onSelect={handleDrillDown} contentHeight="5em" attrs={{ kind: 'action' }}>
              <Typography variant="caption" color="primary">
                Focus on {selected.graphNode.name}
              </Typography>
            </Slice>
          )}
          {views &&
            views.map((view) => (
              <Slice
                key={view.id}
                contentHeight="5em"
                attrs={{ kind: 'nav' }}
                onSelect={() => {
                  handleRelated(view);
                }}>
                <Typography variant="caption" color="textPrimary">
                  {view.name}
                </Typography>
              </Slice>
            ))}
        </PieMenu>
      </ThemeProvider>
    )
  );
};

```

**Comment by Bikeman868** [comment]
I created this [sandbox](https://codesandbox.io/s/vibrant-mirzakhani-w9l7y?file=/src/App.js) so that you can see the issue for yourself.
If you delete one of the slices in this sandbox then you will see the pie menu double in size.

**Comment by psychobolt** [comment]
Looks related to setting container to `overflow: visible`. https://github.com/psychobolt/react-pie-menu/issues/10 

Without this control to the container, the slices will overflow the edges of the radius. There is a alternative to overflow, https://developer.mozilla.org/en-US/docs/Web/CSS/clip-path. Not too sure if this also clips the contents or just the background. Will have to revisit this clipping behavior, when I get the chance.

**Comment by Bikeman868** [comment]
Yes, removing the `overflow: visible;` style does clip the slices properly, however there is still something very odd happening, because when there are 1-3 slices they are still double the size, you just can't see it because of the clipping. You can tell this is still happening by looking at the way the text wraps (or doesn't wrap) in my sandbox example.

**Comment by psychobolt** [comment]
As as mentioned here: https://github.com/psychobolt/react-pie-menu/issues/10#issuecomment-765818539 

The cases where you don't see an odd slice happen is when slice is very small and the `contentContainer` is the almost same or smaller size as the pie. In reality, the constraint is explained below.

> The scale of the container is double the size in for all slices, for __positioning__ purposes against a transformed rect `item`, we need to scale the `contentContainer` so we can transform its origin 50% 50% (calculation for centering `content`). We have a naive calculating of positioning using some old CSS trick (scale a box 200% larger and have the content scale down). __Unfortunately, even if the container is scaled to the pie, and this is expected, we can't depend on that for space calculation.__ 

Onto the question in how can we have a dependable __regain layout control__ (having to calculate width/height itself with the current implementation is one solution). I left in a idea with css __paths__: https://github.com/psychobolt/react-pie-menu/issues/11#issuecomment-767951244

**Comment by Bikeman868** [comment]
My only goal is to make the menu look OK. Ideally the text would wrap within the slice, but everything is a rectangular box in CSS, so its a question of which is the most acceptable compromise. I experimented with making the overflow visible, and I think that looks better than having the text clipped, but then I discovered that it breaks when there are fewer than 4 slices - which looks a whole lot worse, so I have reverted to clipping the text for now.

My conclusion at this point is that a CSS only solution has limitations that make it unsuitable for my use case, and I need to decide if I can live with the limitation of clipped text, perhaps calculating per slice padding to keep the text within the clipping region.

An SVG based solution is definitely possible. Although SVG is very flexible about shapes and transformations, and can have CSS applied to it for many of the attributes, the challenge with SVG is that you won't be able to put any arbitrary HTML in the slices, they would need to be SVG drawing elements, and I don't think that SVG has a text wrapping mechanism.

One option that I might consider is using your current implementation to draw the slices and respond to mouse events, but draw the text over the top of the pie menu. I could even get really fancy and use SVG to draw the text along an arc that matches the curvature of the pie, but this would mean figuring out the line breaks myself and reducing the radius of the curve by the line height for successive lines of text which is quite complicated.

**Comment by Bikeman868** [comment]
Would it be possible to change this package so that the `Slice` component passes props to its children that make it possible to draw the SVG correctly. I think it would need to pass the start and end angle (or center and width) of the slice and the inner and outer radius of the pie.
If you made this change I could write a `TextWedge` component that wraps text within a wedge shape, and place it inside the `Slice`. Thanks.

**Comment by psychobolt** [comment]
Do you have an example of what you are trying to do?

**Comment by psychobolt** [comment]
I guess that would require a small change, but it is not difficult. As for `centerRadius` (inner) and `radius` (outer), I think those could be determined from your current config.  e.g.
```js
<PieMenu 
    radius='125px' 
    centerRadius='30px'
    ...
```

I'm not sure if this works, but right now the Slice is getting those startAngle and endAngle props from the Context. It may be possible to use the same approach for the children e.g.

```js
import { getContext } from 'recompose';
import PropTypes from 'prop-types';

import Content from './Content.componet';

getContext({
  startAngle: PropTypes.number,
  endAngle: PropTypes.number,
})(ContentComponent)
```

**Comment by Bikeman868** [comment]
If I write a `TextWedge` component would you like me to contribute it to this repo?
The `centerRadius` and `radius` really need to be passed in as props from the parent. It's ugly to have to write this:
```
<PieMenu radius="120px" centerRadius="30px" centerX="200px" centerY="200px">
  <Slice>
    <TextWedge radius="120px" centerRadius="30px">Action 1 with long enough name to wrap</TextWedge>
  </Slice>
  <Slice>
    <TextWedge radius="120px" centerRadius="30px">Action 2 with long enough name to wrap</TextWedge>
  </Slice>
</PieMenu>
```

**Comment by psychobolt** [comment]
Were you able to pass the angle props to your `TextWedge` component? If that's the case, then I think we could add centerRadius to the Item's context. Preferably, I would like to change `Item` to use paths and get rid of `ContentContainer`. If you are to contribute `TextWedge`, then please make a branch with your changes ahead of tag [0.2.5](https://github.com/psychobolt/react-pie-menu/tree/react-pie-menu%400.2.5) and I'll take a look at the PR.

**Comment by Bikeman868** [comment]
I finally got to this over the long weekend. After building an SVG based TextWedge component it became apparent that combining text wedges into a pie menu is trivially easy, whereas integrating it into your component is much more involved, so I solved my problem by building a 100% SVG based solution.

**Comment by psychobolt** [comment]
Hi, I just released [v0.3.0](https://github.com/psychobolt/react-pie-menu/tree/react-pie-menu%400.3.0). As mentioned in https://github.com/psychobolt/react-pie-menu/issues/11#issuecomment-775433276 , the context is now managed by the library's theme context (instead of recompose's). Please see [README](https://github.com/psychobolt/react-pie-menu/tree/react-pie-menu%400.3.0#theme-context) for details.
